packages/versioning/src/versioning.ts

```diff
@@ -359,12 +361,113 @@ export function findVersionedNamespace(
   return undefined;
 }
 
+export function $useDependency(
+  context: DecoratorContext,
+  target: EnumMember | Namespace,
+  ...versionRecords: EnumMember[]
+) {
+  const versions: Array<Version> = [];
+  // ensure only valid versions are passed in
+  for (const record of versionRecords) {
+    const ver = checkIsVersion(context.program, record, context.getArgumentTarget(0)!);
+    if (ver) {
+      versions.push(ver);
+    }
+  }
+
+  if (target.kind === "Namespace") {
+    let state = context.program.stateMap(useDependencyNamespaceKey).get(target) as Version[];
+    if (!state) {
+      state = versions;
+    } else {
+      state.push(...versions);
+    }
+    context.program.stateMap(useDependencyNamespaceKey).set(target, state);
+  } else if (target.kind === "EnumMember") {
+    const targetEnum = target.enum;
+    let state = context.program.stateMap(useDependencyEnumKey).get(targetEnum) as Map<
+      EnumMember,
+      Version
+    >;
+    if (!state) {
+      state = new Map<EnumMember, Version>();
+    }
+    for (const v of versions) {
+      state.set(target, v);
+    }
+    context.program.stateMap(useDependencyEnumKey).set(targetEnum, state);
+  }
+}
+
+export function getUseDependencies(
+  program: Program,
+  target: Namespace | Enum,
+  searchEnum: boolean = true
+): Map<Namespace, Map<Version, Version> | Version> | undefined {
+  const result = new Map<Namespace, Map<Version, Version> | Version>();
+  if (target.kind === "Namespace") {
+    let current: Namespace | undefined = target;
+    while (current) {
+      const data = program.stateMap(useDependencyNamespaceKey).get(current) as Version[];
+      if (!data) {
+        // See if the namspace has a version enum
+        if (searchEnum) {
+          const versions = getVersion(program, current)?.getVersions();
+          if (versions?.length) {
+            const enumDeps = getUseDependencies(program, versions[0].enumMember.enum);
+            if (enumDeps) {
+              return enumDeps;
+            }
+          }
+        }
+        current = current.namespace;
+      } else {
+        for (const v of data) {
+          result.set(v.namespace, v);
+        }
+        return result;
+      }
+    }
+    return undefined;
+  } else if (target.kind === "Enum") {
+    const data = program.stateMap(useDependencyEnumKey).get(target) as Map<EnumMember, Version>;
+    if (!data) {
+      return undefined;
+    }
+    const resolved = resolveVersionDependency(program, data);
+    if (resolved instanceof Map) {
+      for (const [enumVer, value] of resolved) {
+        const targetNamespace = value.enumMember.enum.namespace;
+        if (!targetNamespace) {
+          // FIXME: Log diagnostic here???
+          return undefined;
+        }
```

Comment:
@nguerrera this seems to be related to the issue you opened #1520. I could raise a diagnostic here now or just leave a TODO.

Reply:
If the enum isn't inside a namespace then line 440 will yield undefined, so just as in your issue #1520, this logic also will not work if the Version enum isn't inside a namespace.

Reply:
Interesting. I was able to repro #1520 with the enum in a namespace, but a different namespace. So maybe these aren't quite the same? I'm OK with a diagnostic of "must be in a namespace", I think. Could always find a way to relax. It's not as impactful as not being able to share a single enum between two namespaces at all if I understand correctly.

I'd say add the diagnostic here, and comment on #1520 to also consider not having this one either?

Reply:
I tried adding a new diagnostic and another one would consistently get triggered, so I updated this logic to reference that and added a test case.